src/library_dylink.js

```diff
@@ -172,6 +172,16 @@ var LibraryDylink = {
         // correctly.
         rtn.required = true;
       }
+      var value = resolveGlobalSymbol(symName, true).sym;
+      if (!value) {
+        return rtn;
+      }
+      if (typeof value == 'function') {
+        /** @suppress {checkTypes} */
+        rtn.value = addFunction(value, value.sig);
+      } else if (typeof value == {{{ POINTER_JS_TYPE }}}) {
+        rtn.value = value;
+      }
```

Comment:
It looks like this code is very similar to existing code in both `reportUndefinedSymbols` and `updateGOT`.  I wonder if we can factor this better?

Also, I think there is probably some reason we delay updated the GOT at GOTHandler time and instead do it later on.    Perhaps we could instead modify the existing `updateGOT` function?

Reply:
Sounds good I'll try factoring out the shared logic and moving the change to updateGot.

Reply:
> Is __cxa_throw effected because its a JS symbol?

The problem specifically seems to be that `updateGOT` fills in the GOT table for all wasm exports from the loaded library. So it misses JavaScript symbols exported from the main module. Perhaps it would fix it to add GOT entries for all of the JS-implemented library funcs in the main module?

Reply:
How about if we rename `reportUndefinedSymbols` to `handleUndefinedSymbols` and have it take an `reportErrors` flags?

Reply:
Well it seems to me that the only reason we need this GOT update behavior in `reportUndefinedSymbols` is because of JS symbols from the main module which are not handled by `updateGOT`. So they only get added to the GOT if first we have a library that needs them. Why not just stick all JS symbols in the GOT at startup and not worry about having something like `handleUndefinedSymbols`?

Reply:
I don't quite remember the reason for the status quo,  but one reason might be that we want native symbols to take presence over JS symbols.  Another might be that we only want to populate the GOT with symbols that we actually need, not all JS symbols.    I'm just guessing though

Reply:
There's not really any comments in the code about this, or any documentation about when and why it might be necessary to call `handleUndefinedSymbols`. Do we have a test for it? If not, we should make one.

But how do weak linkage symbols work? If I have `libone.so` and it exports a symbol with weak linkage and then `libtwo.so` exports the same symbol again, it doesn't seem like it gets replaced in the GOT. But [Wikipedia says](https://en.wikipedia.org/wiki/Weak_symbol):

> Even if two platforms support the same or similar syntax for marking symbols as weak, the semantics may differ in subtle points, e.g. whether weak symbols during dynamic linking at runtime lose their semantics or not.

So maybe this is compliant. How about just having a set of JS symbols and changing in `GOTHandler` if the current symbol is a JS symbol and if so filling it in.

Reply:
With weak symbols the first library that defines the symbol should win the race, and so I wouldn't expect any replacement in that that.

Reply:
> During linking, a strong symbol can override a weak symbol of the same name. In contrast, in the presence of two strong symbols by the same name, the linker resolves the symbol in favor of the first one found.

So if you see first a weak definition and then later a strong one, shouldn't the second strong one be used over the weak one? They have some examples on the wikipedia page too.
https://en.wikipedia.org/wiki/Weak_symbol#Examples

Reply:
Oh sorry, yes, strong symbols should always win, although I'm not sure how important that edge case is.

Reply:
But you think the case where a JS symbol is overridden by a native symbol is more important? Are the JS symbols automatically counted as weak?

Reply:
No, not necessarily.   I was just trying to imagine/remember the reason for factoring the code this way.

I do think we have good test coverage here,  so I think its reasonably safe to try refactoring this code and if the tests pass it should be fine.

Reply:
There is also a non-zero cost to adding JS functions to the wasm table.. see the `addFunction` helper.

Reply:
Right, in the future where `WebAssembly.Function` is available it will be better. Is it possible we could just import and reexport all these JS functions from the main module so that they would all get added by `updateGOT` in the normal way?